File: decorator/c-sharp/FileDataSource.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Decorator
{
    // Temel işlemlerin tanımlandığı IDataSource arayüzünü uygulayan sınıftır.
    // UML diyagramındaki ConcreteComponent sınıfına denk gelmektedir.
    class FileDataSource : IDataSource
    {
        private string _fileName;

        public FileDataSource(string fileName)
        {
            _fileName = fileName;
        }

        public string GetFileName()
        {
            return _fileName;
        }

        public void ReadData()
        {
            Console.WriteLine($"{_fileName} readed.");
        }

        public void WriteData(object data)
        {
            Console.WriteLine($"data was written to {_fileName}.");
        }
    }
}

File: decorator/go/file-data-source.go
package main

import "fmt"

// fileDataSource UML diyagramındaki ConcreteComponent sınıfına denk gelmektedir.
// UML diyagramındaki ConcreteComponent sınıfına denk gelmektedir.
type fileDataSource struct {
	fileName string
}

func NewFileDataSource(fileName string) *fileDataSource {
	return &fileDataSource{fileName}
}

func (ds fileDataSource) GetFileName() string {
	return ds.fileName
}

func (ds fileDataSource) ReadData() {
	fmt.Println(ds.fileName + " read.")
}

func (ds fileDataSource) WriteData(data []byte) {
	fmt.Println("data was written to " + ds.fileName)
}
